### ISSUE TITLE
limit stream/goroutine creation in Raft transport

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#4422]
via @bdarnell.

From a recent stacktrace:

```
524: select [Created by grpcutil.NewContextWithStopper @ grpc_util.go:96]
   grpcutil     grpc_util.go:91          NewContextWithStopper.func1(#6333, ​*, #2437, *​)
498: select [Created by grpc.(*Server).serveStreams.func1 @ server.go:324]
   transport    transport.go:141         (​*recvBufferReader).Read(*​, *, 0x5, 0x5, 0, 0, 0)
   transport    transport.go:294         (​*Stream).Read(*​, *, 0x5, 0x5, 0, 0, 0)
   io           io.go:298                ReadAtLeast(#6337, ​*, *​, 0x5, 0x5, 0x5, 0, 0, 0)
   io           io.go:316                ReadFull(#6337, ​*, *​, 0x5, 0x5, #6338, 0, 0)
   grpc         rpc_util.go:206          (​*parser).recvMsg(*​, 0, 0, 0, 0, 0, 0)
   grpc         rpc_util.go:288          recv(​*, #6342, #13, *​, 0, 0, #5, *, 0, 0)
   grpc         stream.go:401            (​*serverStream).RecvMsg(*​, #5, *, 0, 0)
   storage      raft.pb.go:166           (​*multiRaftRaftMessageServer).Recv(*​, 0, 0, 0)
   storage      raft_transport.go:88     RaftMessageHandler.RaftMessage(​*, #6340, *​, 0, 0)
   storage      raft.pb.go:147           _MultiRaft_RaftMessage_Handler(#4, ​*, #6339, *​, 0, 0)
   grpc         server.go:583            (​*Server).processStreamingRPC(*​, #6361, ​*, *​, ​*, #8, *​, 0, 0)
   grpc         server.go:658            (​*Server).handleStream(*​, #6361, ​*, *​, *)
   grpc         server.go:323            (​*Server).serveStreams.func1.1(*​, ​*, #6361, *​, *)
```

The first number is the number of goroutines (so 524 and 498, respectively). We should get a handle on those goroutines.

**Comment by tamird** [comment]
Please ping this issue if you see the storage test failure that produces this.

**Comment by tschottdorf** [comment]
This was taken from https://circleci.com/gh/cockroachdb/cockroach/12862

**Comment by bdarnell** [comment]
It looks like a lot of tests are leaking goroutines all the time, but not always enough to cause the test to fail (and they always clean up when the stopper stops). To reproduce the problem you can add a sleep and a log.Fatal to one of `storage/client_*test.go` (I get over 1500 goroutines with a 5s sleep):

```
[bdarnell@rowlf cockroach]$ git diff
diff --git a/storage/client_split_test.go b/storage/client_split_test.go
index 4f5aa35..520a138 100644
--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -878,6 +878,9 @@ func TestSplitSnapshotRace_SnapshotWins(t *testing.T) {
 		t.Fatal(pErr)
 	}
 	mtc.waitForValues(rightKey, splitTimeout, []int64{0, 0, 0, 225, 225, 225})
+
+	time.Sleep(5 * time.Second)
+	log.Fatal("asdf")
 }
 
 // TestStoreSplitReadRace prevents regression of #3148. It begins a couple of
[bdarnell@rowlf cockroach]$ make test PKG=./storage TESTS=SnapshotWins 2>&1 | panicparse
```

**Comment by bdarnell** [comment]
This particular one is a test-only issue: `localRPCTransport` creates a new `RaftMessage` stream for every message and then never closes it. This will be moot if you're going to replace `localRPCTransport` with the real one.

**Comment by tamird** [comment]
Fixed by #4427.
